### PR TITLE
Eliminate Builtin.UnknownObject as an AST type

### PR DESCRIPTION
**Pull request by jrose-apple**
New version of #12325 that doesn't attempt to change the runtime and reflection bits. ~Still needs better comments, more docs, etc.~

### CODE REVIEW

**Comment by jrose-apple** [comment]
apple/swift-lldb#2017
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
apple/swift-lldb#2017
@swift-ci Please test

**Review by slavapestov — APPROVED** [review]
Yay!

File: test/IRGen/dynamic_lookup.sil [hunk 1]
```diff
@@ -60,15 +60,14 @@ bb0(%0 : $AnyObject):
...
 
   // CHECK: [[SEL:%[0-9]+]] = load i8*, i8** @"\01L_selector(f)"
   // CHECK: [[RESPONDS:%[0-9]+]] = load i8*, i8** @"\01L_selector(respondsToSelector:)"
   // CHECK: [[HAS_SEL:%[0-9]]] = call i1 {{.*}}@objc_msgSend {{.*}}(%objc_object* [[OBJECT:%[0-9]+]], i8* [[RESPONDS]], i8* [[SEL]])
   // CHECK: br i1 [[HAS_SEL]]
-  dynamic_method_br %7 : $Builtin.UnknownObject, #X.f!1.foreign, bb1, bb2
+  dynamic_method_br %6 : $@opened("01234567-89ab-cdef-0123-000000000000") AnyObject, #X.f!1.foreign, bb1, bb2
 
-bb1(%8 : $@convention(objc_method) (Builtin.UnknownObject) -> ()):
+bb1(%8 : $@convention(objc_method) (@opened("01234567-89ab-cdef-0123-000000000000") AnyObject) -> ()):
```
Comment (jrose-apple) [inline test/IRGen/dynamic_lookup.sil]: These updates match how `dynamic_method_br` is emitted these days.

File: test/IRGen/enum_value_semantics.sil [hunk 2]
```diff
@@ -76,7 +76,7 @@ enum MultiPayloadGeneric<T> {
 enum MultiPayloadNontrivialSpareBits {
   case payload1(Builtin.NativeObject)
   case payload2(Builtin.Int64)
-  case payload3(Builtin.Int64, Builtin.UnknownObject)
+  case payload3(Builtin.Int64, Builtin.NativeObject)
```
Comment (jrose-apple) [inline test/IRGen/enum_value_semantics.sil]: These tests relied on having spare bits in UnknownObject, which was a questionable assumption. (Good thing we weren't actually using it!)

File: test/IRGen/enum_value_semantics_special_cases_objc.sil [hunk 3]
```diff
@@ -17,86 +17,3 @@ enum NullableObjCRefcounted {
...
-// CHECK:   br label %9
-
-// CHECK: ; <label>:6:
-// CHECK:   %7 = and i64 %0, 4611686018427387903
-// CHECK:   %8 = inttoptr i64 %7 to %objc_object*
-// CHECK:   call void @swift_unknownObjectRelease(%objc_object* %8) #1
-// CHECK:   br label %9
-
-// CHECK: ; <label>:9:
-// CHECK:   ret void
```
Comment (jrose-apple) [inline test/IRGen/enum_value_semantics_special_cases_objc.sil]: These tests also relied on having spare bits in UnknownObject. They don't seem to offer any additional coverage now.

File: test/SILOptimizer/split_critical_edges.sil [hunk 4]
```diff
@@ -90,9 +90,7 @@ lookup1:
   %3 = alloc_box $<τ_0_0> { var τ_0_0 } <Optional<() -> ()>>
   %4 = load %1a : $*AnyObject
   strong_retain %4 : $AnyObject
-  %6 = open_existential_ref %4 : $AnyObject to $@opened("01234567-89ab-cdef-0123-000000000000") AnyObject
-  %7 = unchecked_ref_cast %6 : $@opened("01234567-89ab-cdef-0123-000000000000") AnyObject to $Builtin.UnknownObject
-  dynamic_method_br %7 : $Builtin.UnknownObject, #X.f!1.foreign, bb1, bb2
+  dynamic_method_br %4 : $AnyObject, #X.f!1.foreign, bb1, bb2
```
Comment (jrose-apple) [inline test/SILOptimizer/split_critical_edges.sil]: This one doesn't quite match what we emit today but I figured the test relied on the two branches having the same type.

File: test/SILOptimizer/typed-access-tb-aa.sil [hunk 5]
```diff
@@ -142,17 +142,16 @@ bb0:
 // Now check that a native object address may:
 //
 // 1. May alias raw pointer addresses.
-// 2. Not alias an unknown object. Objective-C objects are allocated differently
-// than Swift objects and may not alias.
+// 2. May alias an unknown object.
```
Comment (jrose-apple) [inline test/SILOptimizer/typed-access-tb-aa.sil]: OOPS GOOD THING WE WEREN'T USING THIS

**Comment by jrose-apple** [comment]
That honestly went better than expected, this should be ready now. I hope it doesn't muck up Instruments or anything though.

(Don't worry, watchers, it'll be tested additionally before it ships in an Xcode.)

apple/swift-lldb#2017
@swift-ci Please test

**Comment by jrose-apple** [comment]
Credit to @varungandhi-apple for getting me to try this again by running into a problem with Builtin.UnknownObject (even if he's already unblocked himself).

**Comment by jrose-apple** [comment]
@atrick or @aschwaighofer, can you check that the test changes I called out above are reasonable?

**Comment by atrick** [comment]
The test case cleanup all looks good.

**Pull request merged**